File: internal/adapters/driven/sns/payment.go
package driven_sns

import (
	"context"
	"encoding/json"
	"errors"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	"github.com/fiap/ms-payment/internal/core/domain/ports"
	infra_aws "github.com/fiap/ms-payment/internal/infra/aws"
)

type PaymentSNS struct {
	client *infra_aws.SNSClient
}

func NewPaymentSNS(client *infra_aws.SNSClient) ports.PaymentTopic {
	return &PaymentSNS{
		client: client,
	}
}

type PaymentTopicType string

const (
	PaymentTopicTypeStatusUpdate    PaymentTopicType = "payment_status_update"
	PaymentTopicTypeCreationSuccess PaymentTopicType = "payment_creation_success"
	PaymentTopicTypeCreationFailure PaymentTopicType = "payment_creation_failure"
)

func (p *PaymentSNS) PublishPaymentStatus(ctx context.Context, externalId string, status string) error {
	input := struct {
		Type       PaymentTopicType `json:"type"`
		ExternalId string           `json:"external_id"`
		Status     string           `json:"status"`
	}{
		Type:       PaymentTopicTypeStatusUpdate,
		ExternalId: externalId,
		Status:     status,
	}

	message, err := json.Marshal(input)
	if err != nil {
		return err
	}

	topicArn, err := p.getTopicArn()
	if err != nil {
		return err
	}

	_, err = p.client.Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(topicArn),
		Message:  aws.String(string(message)),
	})
	if err != nil {
		return err
	}

	return nil
}

func (p *PaymentSNS) PublishPaymentCreationSuccess(ctx context.Context, externalId string, code string) error {
	input := struct {
		Type       PaymentTopicType `json:"type"`
		ExternalId string           `json:"external_id"`
		Code       string           `json:"code"`
	}{
		Type:       PaymentTopicTypeCreationSuccess,
		ExternalId: externalId,
		Code:       code,
	}
	message, err := json.Marshal(input)
	if err != nil {
		return err
	}

	topicArn, err := p.getTopicArn()
	if err != nil {
		return err
	}

	_, err = p.client.Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(topicArn),
		Message:  aws.String(string(message)),
	})

	if err != nil {
		return err
	}

	return nil
}

func (p *PaymentSNS) PublishPaymentCreationFailure(ctx context.Context, externalId string, reason string) error {
	input := struct {
		Type       PaymentTopicType `json:"type"`
		ExternalId string           `json:"external_id"`
		Reason     string           `json:"reason"`
	}{
		Type:       PaymentTopicTypeCreationFailure,
		ExternalId: externalId,
		Reason:     reason,
	}
	message, err := json.Marshal(input)
	if err != nil {
		return err
	}

	topicArn, err := p.getTopicArn()
	if err != nil {
		return err
	}

	_, err = p.client.Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(topicArn),
		Message:  aws.String(string(message)),
	})

	if err != nil {
		return err
	}

	return nil
}

func (p *PaymentSNS) getTopicArn() (string, error) {
	arn := os.Getenv("PAYMENT_TOPIC_ARN")
	if arn == "" {
		return "", errors.New("PAYMENT_TOPIC_ARN environment variable is not set")
	}

	return arn, nil
}

File: src/adapters/driven/sns/payment.ts
import { PublishCommand } from "@aws-sdk/client-sns"

import { HTTPInternalServerError } from "@utils/http"
import { IPaymentTopic } from "@ports/payment_topic"
import { SNSClientWrapper } from "@aws/sns_client"

import {
	PaymentCreationFailureMessage,
	PaymentCreationSuccessMessage,
	PaymentStatusUpdateMessage,
	PaymentTopicMessage,
	PaymentTopicType,
} from "./payment.types"

export class PaymentSNS implements IPaymentTopic {
	constructor(private client: SNSClientWrapper) {}

	private getTopicArn(): string {
		const arn = process.env.PAYMENT_TOPIC_ARN
		if (!arn) {
			throw new HTTPInternalServerError("PAYMENT_TOPIC_ARN environment variable is not set")
		}
		return arn
	}

	private async publish(input: PaymentTopicMessage): Promise<void> {
		const message = JSON.stringify(input)
		const topicArn = this.getTopicArn()

		const command = new PublishCommand({
			TopicArn: topicArn,
			Message: message,
		})

		await this.client.send(command)
	}

	async publishPaymentStatus(externalId: string, status: string): Promise<void> {
		const input: PaymentStatusUpdateMessage = {
			type: PaymentTopicType.StatusUpdate,
			external_id: externalId,
			status: status,
		}

		await this.publish(input)
	}

	async publishPaymentCreationSuccess(externalId: string, code: string): Promise<void> {
		const input: PaymentCreationSuccessMessage = {
			type: PaymentTopicType.CreationSuccess,
			external_id: externalId,
			code: code,
		}

		await this.publish(input)
	}

	async publishPaymentCreationFailure(externalId: string, reason: string): Promise<void> {
		const input: PaymentCreationFailureMessage = {
			type: PaymentTopicType.CreationFailure,
			external_id: externalId,
			reason: reason,
		}

		await this.publish(input)
	}
}